release/nightly_tests/dataset/multi_node_train_benchmark.py

```diff
@@ -108,71 +130,128 @@ def crop_and_flip_image_batch(image_batch):
+        ray_dataset = ray.data.read_images(
+            args.data_root,
+            ray_remote_args=read_ray_remote_args,
+        )
     elif args.file_type == "parquet":
-        ray_dataset = ray.data.read_parquet(args.data_root)
+        ray_dataset = ray.data.read_parquet(
+            args.data_root,
+            ray_remote_args=read_ray_remote_args,
+        )
     else:
         raise Exception(f"Unknown file type {args.file_type}")
+    
+    if args.repeat_ds > 1:
+        ray_dataset = ray_dataset.union(*[Dataset.copy(ray_dataset) for _ in range(args.repeat_ds-1)])
+
+    if cache_input_ds:
+        ray_dataset = ray_dataset.materialize()
+
     # 2) Preprocess data by applying transformation with map_batches()
     ray_dataset = ray_dataset.map_batches(crop_and_flip_image_batch)
+    if cache_output_ds:
+        ray_dataset = ray_dataset.materialize()
 
     def train_loop_per_worker():
         it = train.get_dataset_shard("train")
 
         for i in range(args.num_epochs):
+            print(f"Epoch {i} of {args.num_epochs}")
             num_rows = 0
-            start_t = time.time()
             for batch in it.iter_batches(
                 batch_size=args.batch_size,
-                local_shuffle_buffer_size=args.local_shuffle_buffer_size,
+                # local_shuffle_buffer_size=args.local_shuffle_buffer_size,
                 prefetch_batches=10,
             ):
                 num_rows += args.batch_size
-            end_t = time.time()
-            # Record throughput per epoch.
-            epoch_tput = num_rows / (end_t - start_t)
-            train.report({"tput": epoch_tput, "epoch": i})
 
     # 3) Train TorchTrainer on processed data
     options = DataConfig.default_ingest_options()
     options.preserve_order = args.preserve_order
 
+    if prepartition_ds:
+        class PrepartitionCacheDataConfig(DataConfig):
+            """Instead of using streaming_split to split the dataset amongst workers,
+            pre-partition using Dataset.split(), cache the materialized shards,
+            and assign to each worker."""
+            def __init__(
+                self,
+                datasets_to_split: Union[Literal["all"], List[str]] = "all",
+                execution_options: Optional[ExecutionOptions] = None,
+            ):
+                super().__init__(datasets_to_split, execution_options)
+        
+            def configure(
+                self,
+                datasets: Dict[str, Dataset],
+                world_size: int,
+                worker_handles: Optional[List[ActorHandle]],
+                worker_node_ids: Optional[List[NodeIdStr]],
+                **kwargs,
+            ) -> List[Dict[str, DataIterator]]:
+                assert len(datasets) == 1, "This example only handles the simple case"
+
+                # Split the dataset into shards, materializing the results.
+                materialized_shards = datasets["train"].split(
```

Comment:
Hmm one issue here is that we actually want to partition the input dataset, not the output one. But let's focus on stabilizing the other benchmarks first, since they are more similar to each other than this one. We can follow up on the pre-partitioning in a separate PR.